src/dmd/semantic3.d

```diff
@@ -1356,3 +1346,36 @@ private extern(C++) final class Semantic3Visitor : Visitor
         ad.semanticRun = PASS.semantic3done;
     }
 }
+
+private struct FuncDeclSem3
```

Comment:
Maybe you should instead move this to a new module which encapulates running semantic on `FuncDeclaration`, you could also move `functionSemantic` and `functionSemantic3` there too.
```
module dmd.funcsemantic3;

extern(C++) bool functionSemantic(FuncDeclaration fd);
extern(C++) bool functionSemantic3(FuncDeclaration fd);

private struct Semantic3Function // Or whatever you insist on naming it.
{
  FuncDeclaration funcdecl;
}
```

The visitor entry in `semantic3.d` could then just call `functionSemantic3`.

Reply:
Actually, its even worse than I though.

We have the following methods that do "semantic" on a function.
- DsymbolSemanticVisitor::visit(FuncDeclaration)
- DsymbolSemanticVisitor::funcDeclarationSemantic(FuncDeclaration)
- FuncDeclaration::functionSemantic
- FuncDeclaration::functionSemantic3
- Semantic3Visitor::visit(FuncDeclaration)

I revise my suggestion, they should probably go into two separate leaf modules, each with one public entry point.

Reply:
You're saying fold `Semantic3Visitor::visit(FuncDeclaration)` out of https://github.com/dlang/dmd/blob/master/src/dmd/func.d#L391 so that it calls the function directly? Sure, but I think it would be better suited to a later PR. I want to get this refactored to a  comprehendible state so I can fix  14246 with hopefully a bit more finesse than last time it was tried.

I've half a mind to (eventually) move all the semantic analysis modules to a separate directory, `src/dmd` is mighty cluttered.

I think this is ready to be merged, not sure why buildkite is failing.